`Assessment`: Disable override assessment button when the assessment was not changed

canberkanar opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
#### General
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable and you may remove the first checkmark. If you are unsure, please test on the test servers. -->
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
Closes https://github.com/ls1intum/Artemis/issues/4592

### Description
Override Assessment button is disabled until changes are made.

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->
Prerequisites:
- 1 Instructor

1. Log in to Artemis
2. Navigate to Submissions page of a File Upload, Modeling and Programming Exercise
3. Open Assessment
4. Verify that the Override Assessment button is initially disabled and is enabled when a change is made in the assessment.

### Review Progress
<!-- Each Pull Request should be reviewed by at least two other developers. The code as well as the functionality (= manual test) needs to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage
<!-- Please add the test coverages for all changed files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan. -->
<!-- Please use the schema "Branches % | Lines %" -->
<!-- Lines are the main reference but a significantly lower branch percentage can indicate missing edge cases in the tests. -->
<!-- - ExerciseService.java: 85% | 77% -->
<!-- - programming-exercise.component.ts: 13% | 95% -->

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->
![image](https://user-images.githubusercontent.com/67344158/150694653-8f312ef2-2df1-4bf0-a50c-56d3b7d97f88.png)
![image](https://user-images.githubusercontent.com/67344158/150694665-56ff08fd-f6ba-453a-8910-64a4026258e4.png)

akesfeden commented on src/main/webapp/app/exercises/modeling/assess/modeling-assessment-editor/modeling-assessment-editor.component.ts:
```diff
@@ -193,7 +194,6 @@ export class ModelingAssessmentEditorComponent implements OnInit {
             this.alertService.info('modelingAssessmentEditor.messages.lock');
         }
         this.checkPermissions();
-        this.validateFeedback();
```
Why don't we need this validation anymore (or was it not needed in the first place)?

canberkanar replied:
![image](https://user-images.githubusercontent.com/67344158/152235693-594b612a-2eba-4433-a31b-1d772e16bab0.png)
![image](https://user-images.githubusercontent.com/67344158/152235757-21bdcbb9-2713-4d4c-aa23-56edd1cd4392.png)

I believe it was not intended in the first place. The issue originally stated to solve the problem by implementing in accordance to the text exercises, where the override assessment button is already disabled. I checked that on page load even though the validation is not called, variables that are changed by the function are correct anyways. Also the correction round tag call is made before and after the validation so I believe we do not have any functionality loss here.

akesfeden commented on src/main/webapp/app/exercises/programming/assess/code-editor-tutor-assessment-container.component.ts:
```diff
@@ -539,7 +539,6 @@ export class CodeEditorTutorAssessmentContainerComponent implements OnInit, OnDe
         this.unreferencedFeedback = feedbacks.filter((feedbackElement) => feedbackElement.reference == undefined && feedbackElement.type === FeedbackType.MANUAL_UNREFERENCED);
         this.referencedFeedback = feedbacks.filter((feedbackElement) => feedbackElement.reference != undefined && feedbackElement.type === FeedbackType.MANUAL);
         this.onFeedbackLoaded.emit();
-        this.validateFeedback();
```
Same question as above, why don't we need this validation anymore? Is it related to `handleFeedbackCorrectionRoundTag`?

canberkanar replied:
Similar to the explanation above, besides that there is no call to handleFeedbackCorrectionRoundTag on this class.

Bofan-Zhu reviewed (dismissed):
Tested on ts2. Button is correctly disabled and enabled when editing 👍.

coaxial1 commented on src/main/webapp/app/exercises/file-upload/assess/file-upload-assessment.component.ts:
```diff
@@ -393,6 +393,18 @@ export class FileUploadAssessmentComponent implements OnInit, OnDestroy {
+
+    /**
+     * Calculates the total score of the current assessment.
+     * This function originally checked whether the total score is negative
```
Somehow I don't understand this comment. The function "originally" checked that? Looks like it is still doing it ... 🤔

canberkanar replied:
This function was formed by extracting a code snipped from the "validateAssessment()" function of file-upload-assessment-component. This exact function was already present in modeling-assessment-editor-component and since I created a new function with the exact same code of calculateTotalScore found in the  modeling-assessment-editor-component, I have copied the name of the function and it's documentation for consistency.

If the function description (documentation) below is misleading, we can update it on both of these functions.
![image](https://user-images.githubusercontent.com/67344158/154082347-b810e7f8-e52d-4df5-afce-389862552cd0.png)

canberkanar commented:
tested on ts2

dveljacic commented:
Tested on TS2, and I think I found an unexpected behavior or two. 
The PR fixes only the problem of existing assessments. Once there is an assessment, the button is disabled until we change something. And that's only on the first try. If we stay on the same page, no matter how many times we override assessments, it always stays enabled, even if nothing has been changed. 
Also, if you watch the attached video when assessing the submission the first time, once you submit the assessment, the "Override Assessment" button is still enabled. 

https://user-images.githubusercontent.com/94109535/154118061-4077c4c4-89a2-4d27-bf2c-2ed9c444ad61.mp4

canberkanar commented:
> Tested on TS2, and I think I found an unexpected behavior or two. The PR fixes only the problem of existing assessments. Once there is an assessment, the button is disabled until we change something. And that's only on the first try. If we stay on the same page, no matter how many times we override assessments, it always stays enabled, even if nothing has been changed. Also, if you watch the attached video when assessing the submission the first time, once you submit the assessment, the "Override Assessment" button is still enabled.
> 
>  Untitled.mp4

Thank you very much for your feedback. I am actually aware of these behaviors and the reason the implementation was done this way is the issue itself. There, it was suggested that the already existing implementation of the text exercises was good and it could be reused. The above mentioned behavior can also be implemented if they confirm. @florian-glombik 

https://github.com/ls1intum/Artemis/issues/4592

Santia-go commented:
Tested in ts1. The functionality works as the explanation on this PR establishes. The Override Assessment button is disabled by default because no changes are made when you open the first time the Assessment page. However, I noticed that when I made a change and override it, the button is still available. 
That can confuse the tutors or instructors because the first thing that comes to your mind is that the change was not overridden. 
If this observation was not included to be fixed in this PR, I can approve the first part that is explained in the purpose of this PR.

canberkanar commented:
> Tested on TS2, and I think I found an unexpected behavior or two. The PR fixes only the problem of existing assessments. Once there is an assessment, the button is disabled until we change something. And that's only on the first try. If we stay on the same page, no matter how many times we override assessments, it always stays enabled, even if nothing has been changed. Also, if you watch the attached video when assessing the submission the first time, once you submit the assessment, the "Override Assessment" button is still enabled.
> 
>  Untitled.mp4

Can you please recheck? I have addressed the issues you have mentioned.

canberkanar commented:
> 

@Santia-go I have addressed the issues you have brought up. Can you recheck please?

canberkanar commented:
> Tested in ts1. The functionality works as the explanation on this PR establishes. The Override Assessment button is disabled by default because no changes are made when you open the first time the Assessment page. However, I noticed that when I made a change and override it, the button is still available. That can confuse the tutors or instructors because the first thing that comes to your mind is that the change was not overridden. If this observation was not included to be fixed in this PR, I can approve the first part that is explained in the purpose of this PR.

Can you please recheck? I have addressed the issues you have mentioned. @Santia-go

canberkanar commented:
> Tested on TS2, and I think I found an unexpected behavior or two. The PR fixes only the problem of existing assessments. Once there is an assessment, the button is disabled until we change something. And that's only on the first try. If we stay on the same page, no matter how many times we override assessments, it always stays enabled, even if nothing has been changed. Also, if you watch the attached video when assessing the submission the first time, once you submit the assessment, the "Override Assessment" button is still enabled.
> 
>  Untitled.mp4

Can you please recheck? I have addressed the issues you have mentioned. @dveljacic

willergeorg reviewed (approved):
Tested on TS1 and followed your testing steps for a modeling, text and programming exercise. Works as expected. The issues @dveljacic and @Santia-go mentioned seem to be solved as I was not able to reproduce them.

dveljacic reviewed (approved):
Manually tested on TS2. The "Override Assessment" button is now really intuitive and its behavior can be seen in the attached recording. All problems have been fixed. Nice job! ✅

https://user-images.githubusercontent.com/94109535/155389288-604bcde2-ccfb-41aa-b661-8e3f59aabf92.mp4

Santia-go reviewed (approved):
Tested on TS1. Working as expected and the previous issue in my comment before was also fixed.

akesfeden reviewed (approved):
Code changes look okay, I had some questions about removed validation calls and they are answered.

Bofan-Zhu reviewed (approved):
Approved👍
